grace-core/src/Grace/Import/Resolver.hs

```diff
@@ -0,0 +1,207 @@
+    | EnvVarNotFound Text
+    deriving stock Show
+
+instance Exception EnvResolverError where
+    displayException EnvInvalidURI = "Invalid URI"
+    displayException EnvMissingVarName = "Environment variable name is missing"
+    displayException (EnvVarNotFound k) = Text.unpack ("Environment variable not found: " <> k)
+
+{- | A resolver for files.
+
+     This resolver matches URIs with the @file:@ scheme. The resolver takes the
+     path of the URI, tries to read its content and returns it as a result.
+
+     It will fail if the URI has an authority component.
+-}
+fileResolver :: Resolver
+fileResolver = Resolver \case
+    uri@URI.URI{ URI.uriScheme = Just (URI.unRText -> "file") } -> do
+        case URI.uriAuthority uri of
+            Right auth | auth /= emptyAuthority -> throw FileInvalidURI
+            _ -> return ()
+
+        pieces <- case URI.uriPath uri of
+            Nothing -> throw FileMissingPath
+            Just (_, pieces) -> return pieces
+
+        let input = Parser.Path (pathPiecesToFilePath pieces)
+
+        eitherResult <- Except.runExceptT (Parser.parseInput input)
+
+        result <- case eitherResult of
+            Left e -> throw e
+            Right result -> return result
+
+        return (Just result)
+    _ -> return Nothing
+    where
+        pathPiecesToFilePath = foldl' (</>) "/" . map (Text.unpack . URI.unRText) . NonEmpty.toList
+
+-- | Errors raised by `fileResolver`
+data FileResolverError
+    = FileInvalidURI
+    | FileMissingPath
+    deriving stock Show
+
+instance Exception FileResolverError where
+    displayException FileInvalidURI = "Invalid URI"
+    displayException FileMissingPath = "Filepath is missing"
+
+{- | A resolver that uses external programs to resolve the URI.
+
+     This resolver passes the URI as an command line argument to an executable
+     named @grace-resolver-\<scheme\>@. If that fails because the executable
+     cannot be looked up (i.e. it does not exist in @PATH@) the URI is
+     considered unmatched by this resolver and @Nothing@ is returned. On
+     success, the standard output of that process is expected to be the desired
+     expression to be returned as a result.
+
+     It will fail if the URI has an authority component.
+-}
+externalResolver :: Resolver
+externalResolver = Resolver \case
+    uri@URI.URI{ URI.uriScheme = Just scheme } -> handleDoesNotExist do
+        let cmd = "grace-resolver-" <> Text.unpack (URI.unRText scheme)
+
+        let args = [URI.renderStr uri]
+
+        let pc = Process.proc cmd args
+
+        bytes <- Process.readProcessStdout_ pc
+
+        text <- case Text.Lazy.Encoding.decodeUtf8' bytes of
+            Left e -> throw (ExternalFailedDecodeStdout e)
+            Right text -> return (Text.Lazy.toStrict text)
+
+        let input = Parser.Code "(external)" text
+
+        eitherResult <- Except.runExceptT (Parser.parseInput input)
+
+        result <- case eitherResult of
+            Left e -> throw e
```

Comment:
Did you mean to use `throwIO` here?

The main reason I suggest `throwIO` instead of `throw` is that `throwIO` doesn't risk the exception being triggered accidentally by evaluation.  Or in other words:

```haskell
throw e `seq` () = throw e

throwIO e `seq` () = ()
```

Reply:
Since we `safe-exceptions` this is really just an alias for `throwIO`. See [here](https://hackage.haskell.org/package/safe-exceptions-0.1.7.2/docs/Control-Exception-Safe.html#v:throwIO).